state-sync/storage-service/server/src/lib.rs

```diff
@@ -65,6 +72,134 @@ impl Error {
+        })?;
+        let end_version = known_version
+            .checked_add(num_versions_to_fetch)
+            .ok_or_else(|| {
+                Error::UnexpectedErrorEncountered("End version has overflown!".into())
+            })?;
+
+        // Create the storage request
+        let storage_request = match &self.request {
+            StorageServiceRequest::GetNewTransactionOutputsWithProof(_) => {
+                StorageServiceRequest::GetTransactionOutputsWithProof(
+                    TransactionOutputsWithProofRequest {
+                        proof_version: target_version,
+                        start_version,
+                        end_version,
+                    },
+                )
+            }
+            StorageServiceRequest::GetNewTransactionsWithProof(request) => {
+                StorageServiceRequest::GetTransactionsWithProof(TransactionsWithProofRequest {
+                    proof_version: target_version,
+                    start_version,
+                    end_version,
+                    include_events: request.include_events,
+                })
+            }
+            request => unreachable!("Unexpected subscription request: {:?}", request),
+        };
+        Ok(storage_request)
+    }
+
+    /// Returns the highest version known by the peer
+    fn highest_known_version(&self) -> u64 {
+        match &self.request {
+            StorageServiceRequest::GetNewTransactionOutputsWithProof(request) => {
+                request.known_version
+            }
+            StorageServiceRequest::GetNewTransactionsWithProof(request) => request.known_version,
+            request => unreachable!("Unexpected subscription request: {:?}", request),
+        }
+    }
+
+    /// Returns the highest epoch known by the peer
+    fn highest_known_epoch(&self) -> u64 {
+        match &self.request {
+            StorageServiceRequest::GetNewTransactionOutputsWithProof(request) => {
+                request.known_epoch
+            }
+            StorageServiceRequest::GetNewTransactionsWithProof(request) => request.known_epoch,
+            request => unreachable!("Unexpected subscription request: {:?}", request),
+        }
+    }
+
+    /// Returns the maximum chunk size for the request depending
+    /// on the request type.
+    fn max_chunk_size_for_request(&self, config: StorageServiceConfig) -> u64 {
+        match &self.request {
+            StorageServiceRequest::GetNewTransactionOutputsWithProof(_) => {
+                config.max_transaction_output_chunk_size
+            }
+            StorageServiceRequest::GetNewTransactionsWithProof(_) => {
+                config.max_transaction_chunk_size
+            }
+            request => unreachable!("Unexpected subscription request: {:?}", request),
+        }
+    }
+
+    /// Returns true iff the request has timed out
+    fn timed_out(&self, timeout_ms: u64) -> bool {
```

Comment:
this is not really a timeout but the subscription period? timeout typically refers to when the clients don't receive any response and close the connection

Reply:
I'm not sure there's a hard and fast rule (e.g., I've seen it used both ways 😄). But, I've updated it to be `is_expired`.